controllers/datamovement_controller.go

```diff
@@ -192,62 +218,175 @@ func (r *DataMovementReconciler) Reconcile(ctx context.Context, req ctrl.Request
+	cmdStr, cmdArgs := getCmdAndArgs(dmCmd, numProcesses, dcpProgressInterval, hosts, dm)
+	cmd := exec.CommandContext(ctxCancel, cmdStr, cmdArgs...)
+
 	// Record the start of the data movement operation
 	now := metav1.NowMicro()
 	dm.Status.StartTime = &now
 	dm.Status.State = nnfv1alpha1.DataMovementConditionTypeRunning
+	dm.Status.CommandStatus = &nnfv1alpha1.NnfDataMovementCommandStatus{}
+	cmdStatus := nnfv1alpha1.NnfDataMovementCommandStatus{}
+	cmdStatus.Command = cmd.String()
+	log.Info("Running Command", "cmd", cmdStatus.Command)
 
 	if err := r.Status().Update(ctx, dm); err != nil {
 		return ctrl.Result{}, err
 	}
 
-	// Expand the context with cancel and store it in the map so the cancel
-	// function can be used in another reconciler loop. Also add NamespacedName
-	// so we can retrieve the resource.
-	ctxCancel, cancel := context.WithCancel(ctx)
-	r.contexts.Store(dm.Name, dataMovementCancelContext{
-		ctx:    ctxCancel,
-		cancel: cancel,
-	})
-
 	// Execute the go routine to perform the data movement
 	go func() {
-		var cmd *exec.Cmd
+		// Use a MultiWriter so that we can parse the output and save the full output at the end
+		var combinedOutBuf, parseBuf bytes.Buffer
+		cmd.Stdout = io.MultiWriter(os.Stdout, &parseBuf, &combinedOutBuf)
+		cmd.Stderr = cmd.Stdout // Combine stderr/stdout
+
+		// Use channels to sync progress collection and cmd.Wait().
+		chCommandDone := make(chan bool, 1)
+		chProgressDone := make(chan bool)
+
+		// Start the data movement command
+		cmd.Start()
+
+		// While the command is running, capture and process the output. Read lines until EOF to
+		// ensure we have the latest output. Then use the last regex match to obtain the most recent
+		// progress.
+		go func() {
+			// If progress collection interval is >= 1s, otherwise skip this if
+			if !progressCollectionEnabled(progressCollectInterval) {
+				log.Info("Skipping progress collection - collection interval is less than 1s", "collectInterval", progressCollectInterval)
+				return
+			}
 
-		// TODO: UserId/GroupId
+			var progressInt32 *int32
+			var elapsed metav1.Duration
+			elapsed.Duration = 0
+			progressStart := metav1.NowMicro()
+
+			// Perform the actual collection and update logic
+			parseAndUpdateProgress := func() {
+				progressOutput := ""
+
+				// Read all lines of output until EOF
+				for {
+					line, err := parseBuf.ReadString('\n')
+					if err == io.EOF {
+						break
+					} else if err != nil {
+						log.Error(err, "failed to read progress output")
+					}
+					progressOutput += line
+				}
+
+				// Get the times now before we start the parsing
+				progressNow := metav1.NowMicro()
+				elapsed.Duration = progressNow.Time.Sub(progressStart.Time)
+
+				// Find all matching progress matches but we only care about the
+				// last (most recent one)
+				groups := progressRe.FindAllStringSubmatch(progressOutput, -1)
+				if len(groups) > 0 {
+					mostRecent := groups[len(groups)-1]
```

Comment:
If the last message doesn't have a progress regex, then you're skipping it (since `len(groups) == 0`). That means if dcp stalls doing something other than progressing, we'll be unaware.

Reply:
Agreed, but I thought you didn't want lots of output in `lastMessage`: https://github.com/NearNodeFlash/nnf-dm/pull/60#discussion_r998447214. But that's what you'll get if you don't use the regex match.

Reply:
Only because you're accumulating lines in `progressOutput`. If you move `FindAllSringSubmatch` to parse each line, then you'll get the latest percent (if any). And then the last line is just the last line you read from `parseBuf.ReadString('\n')` (if any)

Reply:
Ok, I've re-written this to parse progress line by line. If we have progress it'll update that, but we'll always update lastMessage with the most recent line and update the timing as well.